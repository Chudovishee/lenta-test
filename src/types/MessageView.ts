export enum MessageAlign {
  left = 'left',
  right = 'right'
}

export default MessageAlign;
