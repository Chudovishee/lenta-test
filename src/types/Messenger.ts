export enum Operator {
  lt = 'lt',
  gt = 'gt'
}

export interface MessageDTO {
  id: number;
  author: string;
  created: number;
  text: string;
}

export interface DialogDTO {
  id: number;
  subject: string;
  created: number;
  lastMessage?: MessageDTO;
}

export interface TimeWhere {
  operator: Operator;
  value: number;
}
