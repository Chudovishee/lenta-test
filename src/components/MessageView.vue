<template>
  <div :class="classes">
    <div class="message-view__text">
      {{ message.text }}
    </div>

    <div
      v-if="!stucked"
      class="message-view__footer"
    >
      <div class="message-view__author">
        {{ message.author }}
      </div>
      <div class="message-view__date">
        {{ date }}
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { MessageDTO } from '@/types/Messenger';
import { MessageAlign } from '@/types/MessageView';

@Component
export default class MessageView extends Vue {
  @Prop({
    type: Object,
    required: true,
  })
  message!: MessageDTO;

  @Prop({
    type: String,
    default: MessageAlign.left,
  })
  align!: MessageAlign;

  @Prop({
    type: Boolean,
    default: false,
  })
  stucked!: boolean;

  get classes(): string[] {
    return [
      'message-view',
      this.align === MessageAlign.right ? 'right' : 'left',
    ];
  }

  get date(): string {
    return new Date(this.message.created)
      .toLocaleString();
  }
}
</script>

<style scoped lang="scss">
.message-view {
  &__text {
    word-break: break-word;
    padding: 24px 32px;
    border-radius: 8px;
  }

  &__footer {
    display: flex;
    margin-top: 8px;
  }

  &__author {
    margin-right: 8px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;

    font-weight: bold;
  }

  &__date {
    white-space: nowrap;
    color: #808080;
  }

  &.left {
    margin-right: 30%;

    .message-view__text {
      background-color: #f6f7f9;
      border-bottom-left-radius: 0;
    }

    .message-view__footer {
      justify-content: left;
    }
  }

  &.right {
    margin-left: 30%;

    .message-view__text {
      background-color: #e9f5f5;
      border-bottom-right-radius: 0;
    }

    .message-view__footer {
      justify-content: right;
    }
  }
}
</style>
