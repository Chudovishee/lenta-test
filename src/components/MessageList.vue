<template>
  <div
    class="message-list"
    @scroll="handleScroll"
  >
    <div
      v-for="(message, index) in messages"
      :key="message.id"
      class="message-list__message-container"
      :class="{ first: index === 0 }"
    >
      <message-view
        :message="message"
        :align="renderMeta[index].align"
        :stucked="renderMeta[index].stucked"
      />
    </div>
  </div>
</template>

<script lang="ts">
import {
  Component, Prop, Watch, Vue,
} from 'vue-property-decorator';
import MessageView from '@/components/MessageView.vue';
import { MessageDTO } from '@/types/Messenger';
import { MessageAlign } from '@/types/MessageView';

interface renderMeta {
  align: MessageAlign;
  stucked: boolean;
}

@Component({
  components: {
    MessageView,
  },
})
export default class MessageList extends Vue {
  @Prop({
    type: Array,
    required: true,
  })
  messages!: MessageDTO[];

  get renderMeta(): renderMeta[] {
    return this.messages.reduce((result: renderMeta[], message: MessageDTO, index: number) => {
      if (index === 0) {
        return [{ align: MessageAlign.left, stucked: false }];
      }

      const prevMeta = result[result.length - 1];
      if (message.author === this.messages[index - 1].author) {
        prevMeta.stucked = true;
        result.push({
          align: prevMeta.align,
          stucked: false,
        });
      } else {
        result.push({
          align: prevMeta.align === MessageAlign.left ? MessageAlign.right : MessageAlign.left,
          stucked: false,
        });
      }
      return result;
    }, []);
  }

  get lowerDate(): number | undefined {
    return this.messages[this.messages.length - 1]?.created;
  }

  @Watch('lowerDate', { immediate: true })
  async onDownMessagePushed(): Promise<void> {
    await this.$nextTick();
    this.$el.scrollTop = this.$el.scrollHeight;
  }

  handleScroll(): void {
    if (this.$el.scrollTop === 0) {
      this.$emit('scroll-top');
    }
  }
}
</script>

<style scoped lang="scss">
.message-list {
  width: 100%;
  height: 100%;

  display: flex;
  flex-direction: column;

  overflow-y: auto;

  &__message-container {
    margin: 8px 16px;

    &.first {
      margin-top: auto;
    }
  }
}
</style>
