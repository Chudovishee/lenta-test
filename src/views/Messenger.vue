<template>
  <div class="messenger">
    <loading v-if="loading" />

    <div class="messenger__dialog-list-container">
      <dialog-list
        :messagesCount="messagesCount"
        :dialogs="dialogs"
        :active-dialog-id="dialogId"
        @activate="handleActivate"
      />
    </div>
    <div class="messenger__right-side">
      <div class="messenger__message-list-container">
        <loading v-if="messagesLoading" />
        <message-list
          v-if="dialogId"
          :messages="messages"
          @scroll-top="handleUpMessages"
        />
        <messages-placeholder v-else/>
      </div>

      <input-message
        :pending="postLoading"
        @send="handleSend"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { mapActions, mapGetters } from 'vuex';
import {
  Component, Prop, Watch, Vue,
} from 'vue-property-decorator';

import DialogList from '@/components/DialogList.vue';
import Loading from '@/components/Loading.vue';
import { DialogDTO, MessageDTO } from '@/types/Messenger';
import MessagesPlaceholder from '@/components/MessagesPlaceholder.vue';
import MessageList from '@/components/MessageList.vue';
import InputMessage from '@/components/InputMessage.vue';

@Component({
  components: {
    InputMessage,
    MessageList,
    MessagesPlaceholder,
    Loading,
    DialogList,
  },
  computed: {
    ...mapGetters({
      messagesCount: 'messenger/messagesCount',
      dialogs: 'messenger/dialogs',
      messages: 'messenger/messages',
    }),
  },
  methods: {
    ...mapActions({
      fetchDialogs: 'messenger/fetchDialogs',
      fetchMessagesCount: 'messenger/fetchMessagesCount',
      fetchMessages: 'messenger/fetchMessages',
      fetchUpMessages: 'messenger/fetchUpMessages',
      postMessage: 'messenger/postMessage',
    }),
  },
})
export default class Messenger extends Vue {
  @Prop({
    type: Number,
  })
  dialogId!: number;

  loading = false;

  messagesLoading = false;

  postLoading = false;

  fetchDialogs!: () => Promise<void>;

  fetchMessagesCount!: () => Promise<void>;

  fetchMessages!: (dialogId: number) => Promise<void>;

  fetchUpMessages!: (dialogId: number) => Promise<void>;

  postMessage!: ({ dialogId, message }: { dialogId: number, message: string }) => Promise<void>;

  messagesCount!: number;

  dialogs!: DialogDTO[];

  messages!: MessageDTO[];

  mounted(): void {
    this.fetchInit();
  }

  @Watch('dialogId')
  async onDialogHandler(dialogId: number): Promise<void> {
    if (dialogId) {
      this.messagesLoading = true;
      await this.fetchMessages(this.dialogId);
      this.messagesLoading = false;
    }
  }

  async fetchInit(): Promise<void> {
    this.loading = true;
    await Promise.all([
      this.fetchMessagesCount(),
      this.fetchDialogs(),
    ]);
    await this.onDialogHandler(this.dialogId);
    this.loading = false;
  }

  handleActivate(dialog: DialogDTO): void {
    this.$router.push({
      name: this.$route.name || undefined,
      params: { dialogId: dialog.id.toString() },
    });
  }

  handleUpMessages(): void {
    this.fetchUpMessages(this.dialogId);
  }

  async handleSend(message: string): Promise<void> {
    this.postLoading = true;
    await this.postMessage({ dialogId: this.dialogId, message });
    this.postLoading = false;
  }
}
</script>

<style scoped lang="scss">
.messenger {
  width: 100%;
  height: 100%;

  display: flex;

  &__dialog-list-container {
    width: 300px;
    flex: 0 0 300px;
  }

  &__right-side {
    flex: 1 0;
    display: flex;
    flex-direction: column;
  }

  &__message-list-container {
    position: relative;
    flex: 1 0;
    overflow: hidden;
  }
}
</style>
