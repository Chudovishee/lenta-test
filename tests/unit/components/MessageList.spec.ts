import {
  createLocalVue, Wrapper, ShallowMountOptions, shallowMount,
} from '@vue/test-utils';

import MessageList from '@/components/MessageList.vue';
import { MessageDTO } from '@/types/Messenger';
import { MessageAlign } from '@/types/MessageView';

const localVue = createLocalVue();
function defaultMount(options: ShallowMountOptions<Vue>): Wrapper<typeof MessageList.prototype> {
  return shallowMount<typeof MessageList.prototype>(MessageList, {
    localVue,
    ...options,
  });
}

function defaultMessage(data?: Partial<MessageDTO>): MessageDTO {
  return {
    id: 1,
    author: 'duck',
    created: 1635070814253,
    text: 'duck puck',
    ...data,
  };
}

describe('MessageList component', () => {
  it('can render with empty messages', () => {
    const component = defaultMount({
      propsData: {
        messages: [],
      },
    });
    expect(component.element).toMatchSnapshot();
  });

  it('can render with messages', () => {
    const component = defaultMount({
      propsData: {
        messages: [
          defaultMessage({ id: 1 }),
          defaultMessage({ id: 2 }),
          defaultMessage({ id: 3 }),
          defaultMessage({ id: 4 }),
        ],
      },
    });
    expect(component.element).toMatchSnapshot();
  });

  describe('renderMeta getter', () => {
    it('can work with empty messages', () => {
      const component = defaultMount({
        propsData: {
          messages: [],
        },
      });

      expect(component.vm.renderMeta).toEqual([]);
    });

    it('can work with one message', () => {
      const component = defaultMount({
        propsData: {
          messages: [
            defaultMessage(),
          ],
        },
      });

      expect(component.vm.renderMeta).toEqual([
        { align: MessageAlign.left, stucked: false },
      ]);
    });

    it('can stuck messages', () => {
      const component = defaultMount({
        propsData: {
          messages: [
            defaultMessage({ id: 1 }),
            defaultMessage({ id: 2 }),
            defaultMessage({ id: 3 }),
            defaultMessage({ id: 4 }),
          ],
        },
      });

      expect(component.vm.renderMeta).toEqual([
        { align: MessageAlign.left, stucked: true },
        { align: MessageAlign.left, stucked: true },
        { align: MessageAlign.left, stucked: true },
        { align: MessageAlign.left, stucked: false },
      ]);
    });

    it('can align messages', () => {
      const component = defaultMount({
        propsData: {
          messages: [
            defaultMessage({ id: 1, author: 'duck' }),
            defaultMessage({ id: 2, author: 'duck' }),
            defaultMessage({ id: 3, author: 'puck' }),
            defaultMessage({ id: 4, author: 'puck' }),
            defaultMessage({ id: 5, author: 'duck' }),
          ],
        },
      });

      expect(component.vm.renderMeta).toEqual([
        { align: MessageAlign.left, stucked: true },
        { align: MessageAlign.left, stucked: false },
        { align: MessageAlign.right, stucked: true },
        { align: MessageAlign.right, stucked: false },
        { align: MessageAlign.left, stucked: false },
      ]);
    });
  });
});
