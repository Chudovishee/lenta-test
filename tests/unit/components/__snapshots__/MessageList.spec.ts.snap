// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList component can render with empty messages 1`] = `
<div
  class="message-list"
/>
`;

exports[`MessageList component can render with messages 1`] = `
<div
  class="message-list"
>
  <div
    class="message-list__message-container first"
  >
    <message-view-stub
      align="left"
      message="[object Object]"
      stucked="true"
    />
  </div>
  <div
    class="message-list__message-container"
  >
    <message-view-stub
      align="left"
      message="[object Object]"
      stucked="true"
    />
  </div>
  <div
    class="message-list__message-container"
  >
    <message-view-stub
      align="left"
      message="[object Object]"
      stucked="true"
    />
  </div>
  <div
    class="message-list__message-container"
  >
    <message-view-stub
      align="left"
      message="[object Object]"
    />
  </div>
</div>
`;
